File: scala-java/src/main/scala/at/twinformatics/scalademo/scala/TailRecursion.scala
package at.twinformatics.scalademo.scala

import scala.annotation.tailrec

/**
  * Example for a classical tail-recursive function. Scala is able to compile a
  * tail-recursive function into an iterative version.
  */
object TailRecursion {

    // @tailrec // Won't compile into tailrec version.
    def fibrec(n: Int): Int = n match {
        case 0 | 1 => n
        case _ => fibrec(n - 1) + fibrec(n - 2)
    }

    @tailrec
    def fib(x: Long, prev: Long = 0, next: Long = 1): Long = x match {
        case 0 => prev
        case 1 => next
        case _ => fib(x - 1, next, next + prev)
    }

    def main(args: Array[String]): Unit = {
        println(fibrec(10))
        println(fib(10))
    }

}

File: scala-java/src/main/java/at/twinformatics/scalademo/java/TailRecursion.java
package at.twinformatics.scalademo.java;

public class TailRecursion {

    // Decompiled version of Scala tailrec version.
    public static long fib(long x, long prev, long next) {
        while(true) {
            long var8;
            if (0L == x) {
                var8 = prev;
            } else {
                if (1L != x) {
                    long var10000 = x - 1L;
                    long var10001 = next;
                    next += prev;
                    prev = var10001;
                    x = var10000;
                    continue;
                }

                var8 = next;
            }

            return var8;
        }
    }

    public static void main(final String[] args) {
        System.out.println(fib(10, 0, 1));
    }

}
